Rules/ProvideDefaultParameterValue.cs

```diff
@@ -0,0 +1,126 @@
+﻿//
+// Copyright (c) Microsoft Corporation.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.Management.Automation.Language;
+using Microsoft.Windows.Powershell.ScriptAnalyzer.Generic;
+using System.ComponentModel.Composition;
+using System.Globalization;
+
+namespace Microsoft.Windows.Powershell.ScriptAnalyzer.BuiltinRules
+{
+    /// <summary>
+    /// ProvideDefaultParameterValue: Check if any uninitialized variable is used.
+    /// </summary>
+    [Export(typeof(IScriptRule))]
+    public class ProvideDefaultParameterValue : IScriptRule
+    {
+        /// <summary>
+        /// AnalyzeScript: Check if any uninitialized variable is used.
+        /// </summary>
+        public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
+        {
+            if (ast == null) throw new ArgumentNullException(Strings.NullAstErrorMessage);
+
+            // Finds all functionAst
+            IEnumerable<Ast> functionAsts = ast.FindAll(testAst => testAst is FunctionDefinitionAst, true);
+
+            foreach (FunctionDefinitionAst funcAst in functionAsts)
+            {
+                // Finds all ParamAsts.
+                IEnumerable<Ast> varAsts = funcAst.FindAll(testAst => testAst is VariableExpressionAst, true);
+
+                // Iterrates all ParamAsts and check if their names are on the list.
+
+                HashSet<string> paramVariables = new HashSet<string>();
+                // only raise the rules for variables in the param block.
+                if (funcAst.Body != null && funcAst.Body.ParamBlock != null && funcAst.Body.ParamBlock.Parameters != null)
+                {
+                    foreach (var paramAst in funcAst.Body.ParamBlock.Parameters)
+                    {
+                        if (Helper.Instance.IsUninitialized(paramAst.Name, funcAst))
```

Comment:
We should add the check for DSC file (that we should omit in the other file) here. Variables in the parameter block of get-targetresource, set-targetresource and test-targetresource should not be considered when running this rule.